{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# 2021 - ENCODE GSE26320 peaks coverage.ipynb\n",
    "\n",
    "\n",
    "See logbook: https://docs.google.com/document/d/1hjkhjo6cB4_AykuQVSDTU07IPD0t6988ZfXwmBb8q0A/edit#\n",
    "\n",
    "0. Process peaks\n",
    "\n",
    "```\n",
    "cd /mnt/stripe/shpynov/chipseq-smk-pipeline\n",
    "\n",
    "# MACS2 narrow\n",
    "for FDR in 0.1 0.05 0.01 1e-3 1e-4 1e-6 1e-8 1e-10; do\n",
    "    snakemake all --use-conda --cores all --directory /mnt/stripe/shpynov/2020_GSE26320/ \\\n",
    "    --config fastq_ext=fastq.gz fastq_dir=/mnt/stripe/shpynov/2020_GSE26320/fastq genome=hg19 \\\n",
    "    macs2_mode=narrow macs2_params=\"-q $FDR\" macs2_suffix=\"q$FDR\" \\\n",
    "    span_fdr=$FDR sicer_fdr=$FDR --rerun-incomplete;\n",
    "done;\n",
    "\n",
    "# MACS2 broad\n",
    "for FDR in 0.1 0.05 0.01 1e-3 1e-4 1e-6 1e-8 1e-10; do\n",
    "    snakemake all --use-conda --cores all --directory /mnt/stripe/shpynov/2020_GSE26320/ \\\n",
    "    --config fastq_ext=fastq.gz fastq_dir=/mnt/stripe/shpynov/2020_GSE26320/fastq genome=hg19 \\\n",
    "    macs2_mode=broad macs2_params=\"--broad --broad-cutoff $FDR\" macs2_suffix=\"broad$FDR\" \\\n",
    "    span_fdr=$FDR sicer_fdr=$FDR --rerun-incomplete;\n",
    "done;\n",
    "\n",
    "```\n",
    "\n",
    "1. Estimate fragment size from MACS2 output = `241.381`\n",
    "\n",
    "```\n",
    "T=$'\\t'\n",
    "for F in macs2/*q0.05*.xls; do \n",
    "    NAME=$(basename $F | sed 's/_q.*//g'); D=$(cat $F | grep '# d =' | sed 's/.*= //g'); \n",
    "    echo \"$NAME$T$D\" >> fragment.tsv; \n",
    "done\n",
    "cat fragment.tsv | awk '{D+=$2; N+=1} END {print(D/N)}'\n",
    "```\n",
    "\n",
    "2. Compute tags for coverage using `SHIFT = 125`\n",
    "\n",
    "```\n",
    "FOLDER=/mnt/stripe/shpynov/2021_GSE26320\n",
    "cd $FOLDER\n",
    "\n",
    "# Create BAM tags to compute coverage\n",
    "cd bam\n",
    "SHIFT=125\n",
    "for BAM in *.bam; do\n",
    "   echo ${BAM};\n",
    "   bedtools bamtobed -i ${BAM} |\\\n",
    "      awk -v OFS='\\t' -v S=${SHIFT} \\\n",
    "      '{if ($6 != \"-\") {print($1, $2+S, $2+S+1)} else {if ($3-S>=1) {print($1, $3-S, $3-S+1)}}}' |\\\n",
    "      sort -u -k1,1 -k3,3n -k2,2n > ${BAM/.bam/.tags}\n",
    "done\n",
    "mkdir ../tags\n",
    "mv *.tags ../tags\n",
    "\n",
    "```\n",
    "3. Compute coverage in peaks\n",
    "\n",
    "```\n",
    "# Compute bed4 files\n",
    "cd ../peaks\n",
    "for F in *0.05*.peak; do\n",
    "    echo $F;\n",
    "    cat $F | awk -v OFS='\\t' '{print $1,$2,$3,$5}' | sort -k1,1 -k2,2n > $F.bed4;\n",
    "done;\n",
    "mkdir ../bed4\n",
    "mv *.bed4 ../bed4/\n",
    "\n",
    "# Add coverage information to peaks\n",
    "mkdir ../covs\n",
    "cd ../bed4\n",
    "CELLS=$(ls *.bed4 | sed -E 's/GSM[0-9]+_//g' | sed 's/_.*//g' | sort --unique);\n",
    "for C in ${CELLS[@]}; do\n",
    "   echo $C;\n",
    "   for M in H3K27ac H3K27me3 H3K36me3 H3K4me1 H3K4me3; do\n",
    "       for R in rep1 rep2; do\n",
    "          echo $R;\n",
    "          for F in *${C}_${M}_${R}*.bed4; do\n",
    "             echo $F;\n",
    "             echo $(ls ../tags/*${C}_${M}_${R}*.tags);\n",
    "             echo $(ls ../tags/*${C}_Input_${R}*.tags);\n",
    "             bedtools intersect -a $F -b ../tags/*${C}_${M}_${R}*.tags -wa -c > ../covs/${F}t; \n",
    "             bedtools intersect -a ../covs/${F}t -b ../tags/*${C}_Input_${R}*.tags -wa -c > ../covs/${F}tc; \n",
    "             rm ../covs/${F}t; \n",
    "          done\n",
    "       done\n",
    "   done\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-03-07T11:38:12.254230Z",
     "start_time": "2019-03-07T11:38:04.952806Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format='retina'\n",
    "\n",
    "from IPython.display import display\n",
    "import pandas as pd\n",
    "\n",
    "import seaborn as sns\n",
    "sns.set_style(\"whitegrid\")\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import os, re\n",
    "import glob\n",
    "from tqdm.auto import tqdm\n",
    "import subprocess, tempfile\n",
    "from itertools import product\n",
    "import plotly.graph_objects as go\n",
    "import plotly.express as px"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bedl(file):\n",
    "    try:\n",
    "        tf = pd.read_csv(file, sep='\\t', header=None)\n",
    "        return tf[2] - tf[1]\n",
    "    except:\n",
    "        return np.zeros(0) # Empty file\n",
    "\n",
    "def lines(file):\n",
    "    try:\n",
    "        tf = pd.read_csv(file, sep='\\t', header=None)\n",
    "        return len(tf)\n",
    "    except:\n",
    "        return 0 # Empty file\n",
    "\n",
    "def join(file):\n",
    "    try:\n",
    "        tf = pd.read_csv(file, sep='\\t', header=None)\n",
    "        return ','.join(f'{c}:{s}-{e}' for c, s, e in zip(tf[0], tf[1], tf[2]))\n",
    "    except:\n",
    "        return '' # Empty file\n",
    "\n",
    "def rank_correlation(file):\n",
    "    try:\n",
    "        tf = pd.read_csv(file, sep='\\t', header=None)\n",
    "        return spearmanr(tf[3], tf[7])[0]\n",
    "    except:\n",
    "        return 0 # Empty file\n",
    "    \n",
    "def d(a, b):\n",
    "    return a / b if b != 0 else 0\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Peaks analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PATH = '/mnt/stripe/shpynov/2021_GSE26320'\n",
    "# Don't have H1 H3K27ac rep2\n",
    "# Don't have Huvec H3K4me3 rep1 \n",
    "# Don't have HepG2 H3K4me1 rep2\n",
    "CELLS = ['GM12878', 'HMEC', 'HSMM', 'K562', 'NHEK', 'NHLF']  # Ignore H1, Huvec, HepG2\n",
    "MODIFICATIONS = ['H3K27ac', 'H3K27me3', 'H3K36me3', 'H3K4me1', 'H3K4me3']\n",
    "REPS = ['rep1', 'rep2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_peaks_fdr(path, suffix, fdrs):\n",
    "    df_fdr = pd.DataFrame(columns=['file', 'modification', 'cell', 'replicate', 'fdr', 'peaks', 'avlength'], \n",
    "                      dtype=object)\n",
    "    for f in tqdm(os.listdir(path)):\n",
    "        if suffix not in f:\n",
    "            continue\n",
    "        fdr = next((fdr for fdr in fdrs if fdr in f), None)\n",
    "        cell = next((cc for cc in CELLS if cc in f), None)\n",
    "        mod = next((m for m in MODIFICATIONS if m in f), None)\n",
    "        rep = 'rep1' if 'rep1' in f else 'rep2'\n",
    "        if fdr and cell and rep and mod:\n",
    "            ps, ls = lines(os.path.join(path, f)), bedl(os.path.join(path, f))\n",
    "            avls = 0 if ps == 0 else sum(ls) / ps\n",
    "            df_fdr.loc[len(df_fdr)] = (f, mod, cell, rep, fdr, ps, avls)\n",
    "    return df_fdr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotdf(df, tool, title):\n",
    "    fig = plt.figure(figsize=(3 * len(MODIFICATIONS), 4))\n",
    "    for i, m in enumerate(MODIFICATIONS):\n",
    "        t = df.loc[df['modification'] == m].copy()\n",
    "        t['fdr'] = t['fdr'].astype(float)\n",
    "        # Plot \n",
    "        ax = plt.subplot(1, len(MODIFICATIONS), i + 1)\n",
    "        ax.title.set_text(m)\n",
    "#         sns.boxplot(data=t, x=\"fdr\", y=title, ax = ax)            \n",
    "        sns.barplot(data=t, x=\"fdr\", y=title, ax = ax, capsize=.2, errwidth=2, edgecolor=\"black\")\n",
    "        ax.xaxis.set_tick_params(rotation=90)\n",
    "        ax.set_ylabel(title)\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f'{PATH}/figures/fdr_{tool}_{title}.png', bbox_inches='tight', dpi=300)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! mkdir /mnt/stripe/shpynov/2021_GSE26320/figures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SPAN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SPAN_FDRS = ['0.1', '0.01', '0.05', '0.001', '0.0001', '1e-06', '1e-08', '1e-10']\n",
    "df_fdr_span = load_peaks_fdr('/mnt/stripe/shpynov/2020_GSE26320/span-islands', '.islands', SPAN_FDRS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Peaks')\n",
    "plotdf(df_fdr_span, 'span', 'peaks')\n",
    "\n",
    "print('Average length')\n",
    "plotdf(df_fdr_span, 'span', 'avlength')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Troubleshooting\n",
    "display(df_fdr_span[(df_fdr_span['modification'] == 'H3K27me3') & \n",
    "                    (df_fdr_span['fdr'] == '0.05') & \n",
    "                    (df_fdr_span['peaks'] < 5000)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('SPAN old peaks')\n",
    "df_fdr_span_old = load_peaks_fdr('/mnt/stripe/shpynov/2020_GSE26320/span', '.peak', SPAN_FDRS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Peaks')\n",
    "plotdf(df_fdr_span_old, 'span_old', 'peaks')\n",
    "\n",
    "print('Average length')\n",
    "plotdf(df_fdr_span_old, 'span_old', 'avlength')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MACS narrow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MACS2_FDRS = ['0.1', '0.01', '0.05', '1e-3', '1-e4', '1e-6', '1e-8', '1e-10']\n",
    "df_fdr_macs2 = load_peaks_fdr('/mnt/stripe/shpynov/2020_GSE26320/macs2', '.narrowPeak', MACS2_FDRS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Peaks')\n",
    "plotdf(df_fdr_macs2, 'macs2', 'peaks')\n",
    "\n",
    "print('Average length')\n",
    "plotdf(df_fdr_macs2, 'macs2', 'avlength')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MACS2 broad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MACS2BROAD_FDRS = ['0.1', '0.05', '1e-2', '1e-3', '1e-4', '1e-6', '1e-8', '1e-10']\n",
    "df_fdr_macs2broad = load_peaks_fdr('/mnt/stripe/shpynov/2020_GSE26320/macs2', '.broadPeak', MACS2BROAD_FDRS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Peaks')\n",
    "plotdf(df_fdr_macs2broad, 'macs2broad', 'peaks')\n",
    "\n",
    "print('Average length')\n",
    "plotdf(df_fdr_macs2broad, 'macs2broad', 'avlength')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SICER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SICER_FDRS = ['0.1', '0.05', '0.01', '0.001', '0.0001', '1e-06', '1e-08', '1e-10']\n",
    "df_fdr_sicer = load_peaks_fdr('/mnt/stripe/shpynov/2020_GSE26320/sicer', 'summary-FDR', SICER_FDRS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Peaks')\n",
    "plotdf(df_fdr_sicer, 'sicer', 'peaks')\n",
    "\n",
    "print('Average length')\n",
    "plotdf(df_fdr_sicer, 'sicer', 'avlength')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Coverage analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FILES = os.listdir(os.path.join(PATH, 'covs'))\n",
    "# display(FILES)\n",
    "\n",
    "reps = []\n",
    "tools = []\n",
    "cells = []\n",
    "mods = []\n",
    "files = []\n",
    "\n",
    "for f in FILES:\n",
    "    rep, tool, cell, mod = None, None, None, None\n",
    "    if 'rep1_rep2' in f:\n",
    "        rep = 'rep1_rep2'  # ENCODE replicated peaks support\n",
    "    else:\n",
    "        rep = 'rep1' if 'rep1' in f else 'rep2'\n",
    "    \n",
    "    \n",
    "    if 'rep1_rep2' in f:\n",
    "        tool = 'ENCODE rep'\n",
    "    elif 'ENCODE' in f:\n",
    "        tool = 'ENCODE'\n",
    "    elif '5.islands' in f:\n",
    "        tool = 'SPAN'\n",
    "    elif 'FDR' in f:\n",
    "        tool = 'SICER'\n",
    "    elif 'narrowPeak' in f:\n",
    "        tool = 'MACS2'\n",
    "    elif 'broadPeak' in f:\n",
    "        tool = 'MACS2 broad'\n",
    "\n",
    "    cell = next((cc for cc in CELLS if cc in f), None)\n",
    "    mod = next((m for m in MODIFICATIONS if m in f), None)\n",
    "\n",
    "    if rep and tool and cell and mod:\n",
    "        reps.append(rep)\n",
    "        tools.append(tool)\n",
    "        cells.append(cell)\n",
    "        mods.append(mod)\n",
    "        files.append(f'{PATH}/covs/{f}')\n",
    "        \n",
    "df = pd.DataFrame(dict(cell=cells, modification=mods, rep=reps, file=files, tool=tools))\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks = []\n",
    "lengths = []\n",
    "avlengths = []\n",
    "\n",
    "for f in df['file']: \n",
    "    ps, ls = lines(f), bedl(f)\n",
    "    avls = 0 if ps == 0 else sum(ls) / ps\n",
    "    peaks.append(ps)\n",
    "    lengths.append(sum(ls))\n",
    "    avlengths.append(avls)\n",
    "\n",
    "df['peaks'] = peaks\n",
    "df['length'] = lengths\n",
    "df['average length'] = avlengths\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coverage_df = pd.DataFrame(columns=['cell', 'replicate', 'modification', 'tags'], dtype=object)\n",
    "for c, r in tqdm(product(CELLS, REPS)):\n",
    "    cc = \\\n",
    "        lines(f'{PATH}/tags/' + next((f for f in os.listdir(f'{PATH}/tags/') if f'{c}_Input_{r}' in f), None))    \n",
    "    coverage_df.loc[len(coverage_df)] = (c, r, 'Input', cc)\n",
    "    for m in MODIFICATIONS:\n",
    "        ct = \\\n",
    "            lines(f'{PATH}/tags/' + next((f for f in os.listdir(f'{PATH}/tags/') if f'{c}_{m}_{r}' in f), None))\n",
    "        coverage_df.loc[len(coverage_df)] = (c, r, m, ct)\n",
    "\n",
    "        \n",
    "def libcoveraget(m, c, r):\n",
    "    return coverage_df[(coverage_df['modification'] == m) & \n",
    "                       (coverage_df['cell'] == c) & \n",
    "                      (coverage_df['replicate'] == r)]['tags'].values[0]\n",
    "def libcoveragec(c, r):\n",
    "    return libcoveraget('Input', c, r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Libraries tags x10mln')\n",
    "fig = plt.figure(figsize=(3 * (len(MODIFICATIONS) + 1), 4))\n",
    "for i, m in enumerate(MODIFICATIONS + ['Input']):\n",
    "    t = coverage_df.loc[coverage_df['modification'] == m].copy()\n",
    "    t['name'] = t['cell'] + ' ' + t['replicate']\n",
    "    ax = plt.subplot(1, len(MODIFICATIONS) + 1, i + 1)\n",
    "    ax.title.set_text(m)\n",
    "    sns.barplot(data=t, x='name', y='tags', ax=ax, color='blue')            \n",
    "    ax.set_ylim(bottom = 0, top = 1.8e7)\n",
    "    ax.xaxis.set_tick_params(rotation=90)\n",
    "    ax.set_ylabel('Tags')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(f'{PATH}/figures/libraries.png', bbox_inches='tight', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading detailed peaks information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "! mkdir {PATH}/bed\n",
    "def coverage_t(file):\n",
    "    try:\n",
    "        tf = pd.read_csv(file, sep='\\t', header=None)\n",
    "        return tf[4]\n",
    "    except:\n",
    "        return [] # Empty file\n",
    "\n",
    "def coverage_c(file):\n",
    "    try:\n",
    "        tf = pd.read_csv(file, sep='\\t', header=None)\n",
    "        return tf[5]\n",
    "    except:\n",
    "        return [] # Empty file\n",
    "\n",
    "def coverage_a(file, cell, replicate):\n",
    "    try:\n",
    "        tf = pd.read_csv(file, sep='\\t', header=None)\n",
    "        scale = min(1, libcoveraget(cell, replicate) / libcoveragec(cell, replicate))\n",
    "        return (tf[4] - tf[5] * scale).clip(lower=0)\n",
    "    except:\n",
    "        return [] # Empty file\n",
    "\n",
    "def zip6(v1, v2, v3, v4, v5, arr):\n",
    "    return [(v1, v2, v3, v4, v5, a) for a in arr]\n",
    "\n",
    "\n",
    "df_comparison = pd.DataFrame(columns=[\n",
    "    'modification', 'cell', 'replicate', 'tool1', 'tool2', \n",
    "    'peaks1', 'peaks2', 'length1', 'length2',\n",
    "    'overlap1', 'overlap2', 'overlap1l', 'overlap2l', \n",
    "    'diff1', 'diff2', 'diff1l', 'diff2l',\n",
    "    'cov1t', 'cov1c', 'cov1a', \n",
    "    'cov2t', 'cov2c', 'cov2a',\n",
    "    'overlap1t', 'overlap1c', 'overlap1a',\n",
    "    'overlap2t', 'overlap2c', 'overlap2a',                                      \n",
    "    'diff1t', 'diff1c', 'diff1a',\n",
    "    'diff2t', 'diff2c', 'diff2a'], dtype=object)\n",
    "\n",
    "tools = list(sorted(set(df['tool'])))\n",
    "\n",
    "df_peaks = []\n",
    "for m, c, r in product(MODIFICATIONS, CELLS, REPS):\n",
    "    print(m, c, r)\n",
    "\n",
    "    covtm = libcoveraget(m, c, r) * 1e-6\n",
    "    covcm = libcoveragec(c, r) * 1e-6\n",
    "\n",
    "    for i in tqdm(range(len(tools))):\n",
    "        tool1 = tools[i]\n",
    "        t1 = df.loc[(df['cell'] == c) & (df['modification'] == m) & \n",
    "                    (df['rep'].str.contains(r)) & (df['tool']==tool1)]\n",
    "\n",
    "        file1 = t1['file'].values[0]\n",
    "#         print('Tool1', tool1, file1)\n",
    "\n",
    "        peaks1, length1 = lines(file1), bedl(file1)\n",
    "        df_peaks.extend(zip6(m, c, r, f'{tool1} ({peaks1})', 'length', length1))            \n",
    "        cov1t, cov1c, cov1a = coverage_t(file1), coverage_c(file1), coverage_a(file1, c, r)\n",
    "        df_peaks.extend(zip6(m, c, r, f'coverage {tool1} ({peaks1})', 'signal', \n",
    "                       [co / covtm for co in cov1t]))\n",
    "        df_peaks.extend(zip6(m, c, r, f'coverage {tool1} ({peaks1})', 'control', \n",
    "                       [co / covcm for co in cov1c]))\n",
    "#                 df_peaks.extend(zip6(m, c, r, f'coverage {tool1} ({peaks1})', 'aggregated', cov1a))\n",
    "        df_peaks.extend(zip6(m, c, r, f'coverage {tool1} ({peaks1})', 'signal pbp', \n",
    "                       [d(co / covtm, l * 1e-3) for co, l in zip(cov1t, length1)]))\n",
    "        df_peaks.extend(zip6(m, c, r, f'coverage {tool1} ({peaks1})', 'control pbp', \n",
    "                       [d(co / covcm, l * 1e-3) for co, l in zip(cov1c, length1)]))\n",
    "#                 df_peaks.extend(zip6(m, c, r, f'coverage {tool1} ({peaks1})', 'aggregated pbp', [d(co, l) for co, l in zip(cov1a, length1)]))\n",
    "\n",
    "            \n",
    "        \n",
    "        for j in range(i + 1, len(tools)):\n",
    "            tool2 = tools[j]\n",
    "            t2 = df.loc[(df['cell'] == c) & (df['modification'] == m) & \n",
    "                        (df['rep'].str.contains(r)) & (df['tool']==tool2)]\n",
    "\n",
    "            file2 = t2['file'].values[0]\n",
    "#             print('Tool2', tool2, file2)\n",
    "\n",
    "            peaks2, length2 = lines(file2), bedl(file2)\n",
    "            df_peaks.extend(zip6(m, c, r, f'{tool2} ({peaks2})', 'length', length2))\n",
    "            cov2t, cov2c, cov2a = coverage_t(file2), coverage_c(file2), coverage_a(file2, c, r)\n",
    "            df_peaks.extend(zip6(m, c, r, f'coverage {tool2} ({peaks2})', 'signal', \n",
    "                           [co / covtm for co in cov2t]))\n",
    "            df_peaks.extend(zip6(m, c, r, f'coverage {tool2} ({peaks2})', 'control', \n",
    "                           [co / covcm for co in cov1c]))\n",
    "#                 df_peaks.extend(zip6(m, c, r, f'coverage {tool2} ({peaks2})', 'aggregated', cov2a))\n",
    "            df_peaks.extend(zip6(m, c, r, f'coverage {tool2} ({peaks2})', 'signal pbp', \n",
    "                           [d(co / covtm, l * 1e-3) for co, l in zip(cov2t, length2)]))\n",
    "            df_peaks.extend(zip6(m, c, r, f'coverage {tool2} ({peaks2})', 'control pbp', \n",
    "                           [d(co / covcm, l * 1e-3) for co, l in zip(cov2c, length2)]))\n",
    "#                 df_peaks.extend(zip6(m, c, r, f'coverage {tool2} ({peaks2})', 'aggregated pbp', [d(co, l) for co, l in zip(cov2a, length2)]))\n",
    "\n",
    "\n",
    "            tf = f'{PATH}/bed/{m}_{c}_{r}_overlap_{tool1}_vs_{tool2}.bed'.replace(' ', '_')\n",
    "            !bedtools intersect -a {file1} -b {file2} -wa -u > {tf}\n",
    "            overlap1, overlap1l = lines(tf), bedl(tf)\n",
    "            df_peaks.extend(zip6(m, c, r, f'overlap {tool1} vs {tool2} ({overlap1})', 'length', overlap1l))            \n",
    "            overlap1t, overlap1c, overlap1a = coverage_t(tf), coverage_c(tf), coverage_a(tf, c, r)\n",
    "            df_peaks.extend(zip6(m, c, r, f'overlap {tool1} vs {tool2} ({overlap1})', 'signal', \n",
    "                           [co / covtm for co in overlap1t]))\n",
    "            df_peaks.extend(zip6(m, c, r, f'overlap {tool1} vs {tool2} ({overlap1})', 'control', \n",
    "                           [co / covcm for co in overlap1c]))\n",
    "#                 df_peaks.extend(zip6(m, c, r, f'overlap {tool1} vs {tool2} ({overlap1})', 'aggregated', overlap1a))\n",
    "            df_peaks.extend(zip6(m, c, r, f'overlap {tool1} vs {tool2} ({overlap1})', 'signal pbp', \n",
    "                           [d(co / covtm, l * 1e-3) for co, l in zip(overlap1t, overlap1l)]))\n",
    "            df_peaks.extend(zip6(m, c, r, f'overlap {tool1} vs {tool2} ({overlap1})', 'control pbp', \n",
    "                           [d(co / covcm, l * 1e-3) for co, l in zip(overlap1c, overlap1l)]))\n",
    "#                 df_peaks.extend(zip6(m, c, r, f'overlap {tool1} vs {tool2} ({overlap1})', 'aggregated pbp', [d(co, l) for co, l in zip(overlap1a, overlap1l)]))\n",
    "\n",
    "\n",
    "            tf = f'{PATH}/bed/{m}_{c}_{r}_overlap_{tool2}_vs_{tool1}.bed'.replace(' ', '_')\n",
    "            !bedtools intersect -b {file1} -a {file2} -wa -u > {tf}\n",
    "            overlap2, overlap2l = lines(tf), bedl(tf)\n",
    "            df_peaks.extend(zip6(m, c, r, f'overlap {tool2} vs {tool1} ({overlap2})', 'length', overlap2l))                        \n",
    "            overlap2t, overlap2c, overlap2a = coverage_t(tf), coverage_c(tf), coverage_a(tf, c, r)\n",
    "            df_peaks.extend(zip6(m, c, r, f'overlap {tool2} vs {tool1} ({overlap2})', 'signal', \n",
    "                           [co / covtm for co in overlap2t]))\n",
    "            df_peaks.extend(zip6(m, c, r, f'overlap {tool2} vs {tool1} ({overlap2})', 'control', \n",
    "                           [co / covcm for co in overlap2c]))\n",
    "#                 df_peaks.extend(zip6(m, c, r, f'overlap {tool2} vs {tool1} ({overlap2})', 'aggregated', overlap2a))\n",
    "            df_peaks.extend(zip6(m, c, r, f'overlap {tool2} vs {tool1} ({overlap2})', 'signal pbp', \n",
    "                           [d(co / covtm, l * 1e-3) for co, l in zip(overlap2t, overlap2l)]))\n",
    "            df_peaks.extend(zip6(m, c, r, f'overlap {tool2} vs {tool1} ({overlap2})', 'control pbp', \n",
    "                           [d(co / covcm, l * 1e-3) for co, l in zip(overlap2c, overlap2l)]))\n",
    "#                 df_peaks.extend(zip6(m, c, r, f'overlap {tool2} vs {tool1} ({overlap2})', 'aggregated pbp', [d(co, l) for co, l in zip(overlap2a, overlap2l)]))\n",
    "\n",
    "\n",
    "            tf = f'{PATH}/bed/{m}_{c}_{r}_diff_{tool1}_vs_{tool2}.bed'.replace(' ', '_')\n",
    "            !bedtools intersect -a {file1} -b {file2} -v > {tf}\n",
    "            diff1, diff1l = lines(tf), bedl(tf)\n",
    "            df_peaks.extend(zip6(m, c, r, f'diff {tool1} - {tool2} ({diff1})', 'length', diff1l))                        \n",
    "            diff1t, diff1c, diff1a = coverage_t(tf), coverage_c(tf), coverage_a(tf, c, r)\n",
    "            df_peaks.extend(zip6(m, c, r, f'diff {tool1} - {tool2} ({diff1})', 'signal', \n",
    "                           [co / covtm for co in diff1t]))\n",
    "            df_peaks.extend(zip6(m, c, r, f'diff {tool1} - {tool2} ({diff1})', 'control', \n",
    "                           [co / covcm for co in diff1c]))\n",
    "#                 df_peaks.extend(zip6(m, c, r, f'diff {tool1} - {tool2} ({diff1})', 'aggregated', diff1a))\n",
    "            df_peaks.extend(zip6(m, c, r, f'diff {tool1} - {tool2} ({diff1})', 'signal pbp', \n",
    "                           [d(co / covtm, l * 1e-3) for co, l in zip(diff1t, diff1l)]))\n",
    "            df_peaks.extend(zip6(m, c, r, f'diff {tool1} - {tool2} ({diff1})', 'control pbp', \n",
    "                           [d(co / covcm, l * 1e-3) for co, l in zip(diff1c, diff1l)]))\n",
    "#                 df_peaks.extend(zip6(m, c, r, f'diff {tool1} - {tool2} ({diff1})', 'aggregated pbp', [d(co, l) for co, l in zip(diff1a, diff1l)]))\n",
    "\n",
    "            tf = f'{PATH}/bed/{m}_{c}_{r}_diff_{tool2}_vs_{tool1}.bed'.replace(' ', '_')        \n",
    "            !bedtools intersect -b {file1} -a {file2} -v > {tf}\n",
    "            diff2, diff2l = lines(tf), bedl(tf)\n",
    "            df_peaks.extend(zip6(m, c, r, f'diff {tool2} - {tool1} ({diff2})', 'length', diff2l))            \n",
    "            diff2t, diff2c, diff2a = coverage_t(tf), coverage_c(tf), coverage_a(tf, c, r)\n",
    "            df_peaks.extend(zip6(m, c, r, f'diff {tool2} - {tool1} ({diff2})', 'signal', \n",
    "                           [co / covtm for co in diff2t]))\n",
    "            df_peaks.extend(zip6(m, c, r, f'diff {tool2} - {tool1} ({diff2})', 'control', \n",
    "                           [co / covcm for co in diff2c]))\n",
    "#                 df_peaks.extend(zip6(m, c, r, f'diff {tool2} - {tool1} ({diff2})', 'aggregated', diff2a))\n",
    "            df_peaks.extend(zip6(m, c, r, f'diff {tool2} - {tool1} ({diff2})', 'signal pbp', \n",
    "                           [d(co / covtm, l * 1e-3) for co, l in zip(diff2t, diff2l)]))\n",
    "            df_peaks.extend(zip6(m, c, r, f'diff {tool2} - {tool1} ({diff2})', 'control pbp', \n",
    "                           [d(co / covcm, l * 1e-3) for co, l in zip(diff2c, diff2l)]))\n",
    "#                 df_peaks.extend(zip6(m, c, r, f'diff {tool2} - {tool1} ({diff2})', 'aggregated pbp', [d(co, l) for co, l in zip(diff2a, diff2l)]))\n",
    "\n",
    "            df_comparison.loc[len(df_comparison)] = (\n",
    "                m, c, r, tool1, tool2, \n",
    "                peaks1, peaks2, sum(length1), sum(length2), \n",
    "                overlap1, overlap2, sum(overlap1l), sum(overlap2l),\n",
    "                diff1, diff2, sum(diff1l), sum(diff2l),\n",
    "                sum(cov1t), sum(cov1c), sum(cov1a), \n",
    "                sum(cov2t), sum(cov2c), sum(cov2a), \n",
    "                sum(overlap1t), sum(overlap1c), sum(overlap1a),\n",
    "                sum(overlap2t), sum(overlap2c), sum(overlap2a),\n",
    "                sum(diff1t), sum(diff1c), sum(diff1a),\n",
    "                sum(diff2t), sum(diff2c), sum(diff2a)\n",
    "            )\n",
    "            \n",
    "            \n",
    "df_peaks = pd.DataFrame(df_peaks, columns=['modification', 'cell', 'replicate', 'name', 'type', 'value'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(df_peaks.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(df_comparison.head())\n",
    "df_comparison.to_csv(f'{PATH}/comparison.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Agggregate peaks, length, coverage and density information')\n",
    "df_comparison_average = pd.DataFrame(columns=['m', 'c', 'r', 'coverage', 'tool1', 'tool2', \n",
    "      'peaks1', 'peaks2', 'overlap1', 'overlap2', 'diff1', 'diff2',\n",
    "      'avl1', 'avl2', 'ov_avl1', 'ov_avl2', 'diff_avl1', 'diff_avl2', \n",
    "      'avcov1', 'avcov2', 'ov_avcov1', 'ov_avcov2', 'diff_avcov1', 'diff_avcov2', \n",
    "      'covpbp1', 'covpbp2', 'ov_covpbp1', 'ov_covpbp2', 'diff_covpbp1', 'diff_covpbp2'], dtype=object)\n",
    "\n",
    "for i, row in tqdm(df_comparison.iterrows()):\n",
    "    (m, c, r, tool1, tool2, \n",
    "     peaks1, peaks2, length1, length2, \n",
    "     overlap1, overlap2, overlap1l, overlap2l,\n",
    "     diff1, diff2, diff1l, diff2l,\n",
    "     cov1t, cov1c, cov1a, \n",
    "     cov2t, cov2c, cov2a, \n",
    "     overlap1t, overlap1c, overlap1a,\n",
    "     overlap2t, overlap2c, overlap2a,\n",
    "     diff1t, diff1c, diff1a,\n",
    "     diff2t, diff2c, diff2a) = row    \n",
    "    df_comparison_average.loc[len(df_comparison_average)] = (\n",
    "        m, c, r, 'signal', tool1, tool2, \n",
    "        peaks1, peaks2, overlap1, overlap2, diff1, diff2,\n",
    "        d(length1, peaks1), d(length2, peaks2), d(overlap1l, overlap1), d(overlap2l, overlap2), d(diff1l, diff1), d(diff2l, diff2),  \n",
    "        d(cov1t, peaks1), d(cov2t, peaks2), d(overlap1t, overlap1), d(overlap2t, overlap2), d(diff1t, diff1), d(diff2t, diff2), \n",
    "        d(cov1t, length1), d(cov2t, length2), d(overlap1t, overlap1l), d(overlap2t, overlap2l), d(diff1t, diff1l), d(diff2t, diff2l)\n",
    "      )\n",
    "\n",
    "    df_comparison_average.loc[len(df_comparison_average)] = (\n",
    "        m, c, r, 'control', tool1, tool2, \n",
    "        peaks1, peaks2, overlap1, overlap2, diff1, diff2,\n",
    "        d(length1, peaks1), d(length2, peaks2), d(overlap1l, overlap1), d(overlap2l, overlap2), d(diff1l, diff1), d(diff2l, diff2),  \n",
    "        d(cov1c, peaks1), d(cov2c, peaks2), d(overlap1c, overlap1), d(overlap2c, overlap2), d(diff1c, diff1), d(diff2c, diff2), \n",
    "        d(cov1c, length1), d(cov2c, length2), d(overlap1c, overlap1l), d(overlap2c, overlap2l), d(diff1c, diff1l), d(diff2c, diff2l)\n",
    "      )\n",
    "\n",
    "#     df_comparison_average.loc[len(df_comparison_average)] = (\n",
    "#         m, c, r, 'aggregated', tool1, tool2, \n",
    "#         peaks1, peaks2, overlap1, overlap2, diff1, diff2,\n",
    "#         d(length1, peaks1), d(length2, peaks2), d(overlap1l, overlap1), d(overlap2l, overlap2), d(diff1l, diff1), d(diff2l, diff2),  \n",
    "#         d(cov1a, peaks1), d(cov2a, peaks2), d(overlap1a, overlap1), d(overlap2a, overlap2), d(diff1a, diff1), d(diff2a, diff2), \n",
    "#         d(cov1a, length1), d(cov2a, length2), d(overlap1a, overlap1l), d(overlap2a, overlap2l), d(diff1a, diff1l), d(diff2a, diff2l)\n",
    "#       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option(\"display.max_columns\", None)\n",
    "display(df_comparison_average.head())\n",
    "df_comparison_average.to_csv(f'{PATH}/comparison2.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_comparison_average = pd.read_csv(f'{PATH}/comparison2.csv')\n",
    "# df_comparison_average"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aggregated for all tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "COLS=dict(\n",
    "    peaks=['peaks1', 'peaks2', 'diff1', 'diff2'],\n",
    "    length=['avl1', 'avl2', 'diff_avl1', 'diff_avl2'],\n",
    "#     coverage=['avcov1', 'avcov2', 'diff_avcov1', 'diff_avcov2'],\n",
    "    density=['covpbp1', 'covpbp2', 'diff_covpbp1', 'diff_covpbp2']\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Per modification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_tools_summary_modification(df_comparison, m):\n",
    "\n",
    "    plt.figure(figsize=(5 * len(COLS), 6))\n",
    "\n",
    "    for k, (name, columns) in enumerate(COLS.items()):\n",
    "        types =  ['signal', 'control'] if name == 'coverage' or name == 'density' else ['signal']\n",
    "        tts = []\n",
    "        seen_data = {tt: set([]) for tt in types}\n",
    "        for i in range(len(tools)):\n",
    "            for j in range(i + 1, len(tools)):\n",
    "                tool1 = tools[i]\n",
    "                tool2 = tools[j]\n",
    "\n",
    "                for tt in types:\n",
    "                    ts = df_comparison.loc[(df_comparison['m'] == m) & \n",
    "                                           (df_comparison['tool1'] == tool1) & \n",
    "                                           (df_comparison['tool2'] == tool2) & \n",
    "                                           (df_comparison['coverage'] == tt)].copy()\n",
    "                    ts['exp'] = [f'{c} {r}' for c, r in zip(ts['c'], ts['r'])]\n",
    "                    t = ts[['exp'] + columns].copy()\n",
    "                    t.rename(dict(zip(columns, \n",
    "                                      [tool1, tool2, f'diff {tool1} - {tool2}', f'diff {tool2} - {tool1}'])), \n",
    "                             axis=1, inplace=True)\n",
    "\n",
    "                    t = pd.melt(t, id_vars=['exp'], value_vars=list(t.columns).remove('exp'))\n",
    "                    t['type'] = tt\n",
    "                    t = t.loc[~(t['variable'].isin(seen_data[tt]))].copy()\n",
    "                    seen_data[tt].update(t['variable'])\n",
    "                    tts.append(t)\n",
    "\n",
    "        t = pd.concat(tts)\n",
    "        t.sort_values(by=['variable'], inplace=True)\n",
    "        # Add mean variables\n",
    "        for var in set(t['variable']):\n",
    "            tm = t[t[\"variable\"] == var][\"value\"].mean()\n",
    "            if name == 'peaks' or name == 'length':\n",
    "                t.loc[t['variable'] == var, 'variable'] = f'{var} ({int(tm)})'\n",
    "            else:\n",
    "                t.loc[t['variable'] == var, 'variable'] = f'{var} ({tm:.3f})'\n",
    "\n",
    "                \n",
    "        # Plot \n",
    "        ax = plt.subplot(1, len(COLS), k + 1)\n",
    "        ax.title.set_text(f'{m} {name}')\n",
    "        if name == 'coverage' or name == 'density':\n",
    "            sns.boxplot(data=t, x='variable', y='value', hue='type', ax=ax)\n",
    "        else:\n",
    "            sns.boxplot(data=t, x='variable', y='value', ax=ax)            \n",
    "        ax.xaxis.set_tick_params(rotation=90)\n",
    "        ax.set_ylabel(name)\n",
    "\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f'{PATH}/figures/{m}.png', bbox_inches='tight', dpi=300)        \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for m in MODIFICATIONS:\n",
    "    print(m)\n",
    "    plot_tools_summary_modification(df_comparison_average, m)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Per type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "COLS=dict(\n",
    "    peaks=['peaks1', 'peaks2', 'diff1', 'diff2'],\n",
    "    length=['avl1', 'avl2', 'diff_avl1', 'diff_avl2'],\n",
    "    coverage=['avcov1', 'avcov2', 'diff_avcov1', 'diff_avcov2'],\n",
    "    density=['covpbp1', 'covpbp2', 'diff_covpbp1', 'diff_covpbp2']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_tools_summary_type(df_comparison, title, name):\n",
    "    \"Columns should contain tool1, tool2, diff1, diff2 column names\"\n",
    "    columns = COLS[name]\n",
    "    types =  ['signal', 'control'] if name == 'coverage' or name == 'density' else ['signal']\n",
    "    plt.figure(figsize=(5 * len(MODIFICATIONS), 6))\n",
    "\n",
    "    for k, m in enumerate(MODIFICATIONS):\n",
    "\n",
    "        tts = []\n",
    "        seen_data = {tt: set([]) for tt in types}\n",
    "        for i in range(len(tools)):\n",
    "            for j in range(i + 1, len(tools)):\n",
    "                tool1 = tools[i]\n",
    "                tool2 = tools[j]\n",
    "\n",
    "                for tt in types:\n",
    "                    ts = df_comparison.loc[(df_comparison['m'] == m) & \n",
    "                                           (df_comparison['tool1'] == tool1) & \n",
    "                                           (df_comparison['tool2'] == tool2) & \n",
    "                                           (df_comparison['coverage'] == tt)].copy()\n",
    "                    ts['exp'] = [f'{c} {r}' for c, r in zip(ts['c'], ts['r'])]\n",
    "                    t = ts[['exp'] + columns].copy()\n",
    "                    t.rename(dict(zip(columns, \n",
    "                                      [tool1, tool2, f'diff {tool1} - {tool2}', f'diff {tool2} - {tool1}'])), \n",
    "                             axis=1, inplace=True)\n",
    "\n",
    "                    t = pd.melt(t, id_vars=['exp'], value_vars=list(t.columns).remove('exp'))\n",
    "                    t['type'] = tt\n",
    "                    t = t.loc[~(t['variable'].isin(seen_data[tt]))].copy()\n",
    "                    seen_data[tt].update(t['variable'])\n",
    "                    tts.append(t)\n",
    "\n",
    "        t = pd.concat(tts)\n",
    "        t.sort_values(by=['variable'], inplace=True)\n",
    "        # Add mean variables\n",
    "        for var in set(t['variable']):\n",
    "            tm = t[t[\"variable\"] == var][\"value\"].mean()\n",
    "            if name == 'peaks' or name == 'length':\n",
    "                t.loc[t['variable'] == var, 'variable'] = f'{var} ({int(tm)})'\n",
    "            else:\n",
    "                t.loc[t['variable'] == var, 'variable'] = f'{var} ({tm:.3f})'                \n",
    "        # Plot \n",
    "        ax = plt.subplot(1, len(MODIFICATIONS), k + 1)\n",
    "        ax.title.set_text(m)\n",
    "        if name == 'coverage' or name == 'density':\n",
    "            sns.boxplot(data=t, x='variable', y='value', hue='type', ax=ax)\n",
    "        else:\n",
    "            sns.boxplot(data=t, x='variable', y='value', ax=ax)            \n",
    "        ax.xaxis.set_tick_params(rotation=90)\n",
    "        ax.set_ylabel(title)\n",
    "\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f'{PATH}/figures/{name}.png', bbox_inches='tight', dpi=300)        \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Peaks number')\n",
    "plot_tools_summary_type(df_comparison_average, 'Peaks number', 'peaks')\n",
    "print('Peaks length')\n",
    "plot_tools_summary_type(df_comparison_average, 'Peaks length', 'length')\n",
    "print('Coverage')\n",
    "plot_tools_summary_type(df_comparison_average, 'Coverage', 'coverage')\n",
    "print('Density')\n",
    "plot_tools_summary_type(df_comparison_average, 'Density', 'density')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pairwise peaks number venn diagrams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from matplotlib_venn import venn2, venn2_circles\n",
    "\n",
    "\n",
    "for m in MODIFICATIONS:\n",
    "    print(m)\n",
    "    plt.figure(figsize=(20, 3 * len(CELLS)))\n",
    "    k = 1\n",
    "    tm = df_comparison.loc[(df_comparison['modification'] == m)].copy()\n",
    "    tm.sort_values(by=['cell', 'tool1', 'tool2', 'replicate'], inplace=True)\n",
    "    for _, row in tqdm(tm.iterrows()):\n",
    "        c, r, tool1, tool2, peaks1, peaks2, overlap1, overlap2 = \\\n",
    "            row['cell'], row['replicate'], row['tool1'], row['tool2'], \\\n",
    "            row['peaks1'], row['peaks2'], row['overlap1'], row['overlap2']\n",
    "        if not (tool1 == 'SPAN' or tool2 == 'SPAN'):\n",
    "            continue\n",
    "\n",
    "        # Tweak subsets to avoid empty areas\n",
    "        subsets = (max(0.15 * max(peaks1, peaks2), peaks1 - overlap1),\n",
    "                  max(0.15 * max(peaks1, peaks2), peaks2 - overlap2),\n",
    "                  max(0.3 * max(peaks1, peaks2), overlap1, overlap2))\n",
    "        # Plot \n",
    "        ax = plt.subplot(len(CELLS), 6, k)\n",
    "        k += 1\n",
    "\n",
    "        v = venn2(subsets=subsets,set_labels=(tool1, tool2), set_colors=('r', 'g'), alpha = 0.5, ax=ax)\n",
    "        venn2_circles(subsets=subsets, color='grey')\n",
    "        v.get_label_by_id('10').set_text(str(peaks1 - overlap1))\n",
    "        v.get_label_by_id('01').set_text(str(peaks2 - overlap2))\n",
    "        if overlap1 != 0 and overlap2 != 0:\n",
    "            prop = f'{overlap1/overlap2:.1f} : 1' if overlap1 > overlap2 else f'1 : {overlap2/overlap1:.1f}'\n",
    "            prop = prop.replace('.0', '')  # Cosmetics\n",
    "        else:\n",
    "            prop = ''\n",
    "        v.get_label_by_id('11').set_text(f'{overlap1} : {overlap2}\\n{prop}')\n",
    "        ax.title.set_text(f'{m} {c} {r}')\n",
    "        \n",
    "    plt.tight_layout()    \n",
    "    plt.savefig(f'{PATH}/figures/venn_{m}.png', bbox_inches='tight', dpi=300)        \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Detailed analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_no = df_peaks.loc[~(df_peaks['name'].str.startswith('overlap'))]\n",
    "df_no.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Detailed peaks information for each modification, cell line and replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for m, c, r in tqdm(product(MODIFICATIONS, CELLS, REPS)):\n",
    "    print(m, c, r)\n",
    "    ts = df_no.loc[(df_no['cell'] == c) & (df_no['modification'] == m) & (df_no['replicate'] == r)]\n",
    "    if len(ts) == 0:\n",
    "        continue\n",
    "    \n",
    "    # Plot \n",
    "    plt.figure(figsize=(15, 6))\n",
    "\n",
    "    t = ts.loc[ts['type'] == 'length'][['name', 'value']].copy()\n",
    "    t.sort_values(by=['name'], inplace=True)\n",
    "    t['value'].clip(upper=20000, inplace=True)\n",
    "    ax = plt.subplot(1, 3, 1)\n",
    "    sns.boxplot(data=t, x='name', y='value', ax=ax)\n",
    "    ax.title.set_text(f'{c} {m} {r} Length peaks')\n",
    "    ax.xaxis.set_tick_params(rotation=90)\n",
    "    ax.set_ylabel('Length')\n",
    "\n",
    "    t = ts[(ts['type'] != 'length') & (~(ts['type'].str.contains('pbp')))].copy()\n",
    "    t.sort_values(by=['name', 'type'], inplace=True)\n",
    "    t['value'].clip(upper=50, inplace=True)\n",
    "    ax = plt.subplot(1, 3, 2)\n",
    "    sns.boxplot(data=t, x='name', hue='type', y='value', ax=ax)\n",
    "    ax.title.set_text(f'{c} {m} {r} Coverage in peaks')\n",
    "    ax.xaxis.set_tick_params(rotation=90)\n",
    "    ax.set_ylabel('RPM')\n",
    "\n",
    "\n",
    "    t = ts[ts['type'].str.contains('pbp')].copy()\n",
    "    t.sort_values(by=['name', 'type'], inplace=True)\n",
    "    t['value'].clip(upper=10.0, inplace=True)      \n",
    "    ax = plt.subplot(1, 3, 3)\n",
    "    sns.boxplot(data=t, x='name', hue='type', y='value', ax=ax)\n",
    "    ax.title.set_text(f'{c} {m} {r} Density in peaks')\n",
    "    ax.xaxis.set_tick_params(rotation=90)\n",
    "    ax.set_ylabel('RPKM')\n",
    "\n",
    "    plt.tight_layout()    \n",
    "    plt.savefig(f'{PATH}/figures/peaks_{m}_{c}_{r}.png', bbox_inches='tight', dpi=300)        \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Peaks length and density distributions analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print('Length distribution')\n",
    "for m in MODIFICATIONS:\n",
    "    print(m)\n",
    "    ts = df_no.loc[(df_no['modification'] == m) & (df_no['type'] == 'length')]\n",
    "    plt.figure(figsize=(14, 6))\n",
    "    k = 1\n",
    "    for i in tqdm(range(len(tools))):\n",
    "        for j in range(i + 1, len(tools)):\n",
    "            tool1 = tools[i]\n",
    "            tool2 = tools[j]\n",
    "            if not (tool1 == 'SPAN' or tool2 == 'SPAN'):\n",
    "                continue\n",
    "\n",
    "            tst = ts[(ts['name'].str.startswith(f'{tool1} (')) | \n",
    "                     (ts['name'].str.startswith(f'{tool2} (')) | \n",
    "                     (ts['name'].str.contains(f'coverage {tool1} (', regex=False)) | \n",
    "                     (ts['name'].str.contains(f'coverage {tool2} (', regex=False)) | \n",
    "                     (ts['name'].str.contains(f'diff {tool1} - {tool2} (', regex=False)) | \n",
    "                     (ts['name'].str.contains(f'diff {tool2} - {tool1} (', regex=False))].copy()\n",
    "            tst['name'] = [re.sub(' \\(.*', '', n) for n in tst['name']]\n",
    "\n",
    "            # Plot             \n",
    "            tst.sort_values(by=['name', 'type'], inplace=True)\n",
    "            tst.rename(dict(value='Length'), axis=1, inplace=True)\n",
    "            ax = plt.subplot(2, 3, k)\n",
    "            sns.histplot(data=tst.loc[tst['Length'] <= 10000], x='Length', hue='name', \n",
    "                         stat='density', kde=True, ax=ax)\n",
    "            ax.title.set_text(f'{m} {tool1} vs {tool2} l<=10k')\n",
    "            ax = plt.subplot(2, 3, k + 3)\n",
    "            sns.histplot(data=tst.loc[tst['Length'] > 10000], x='Length', hue='name', \n",
    "                         log_scale=True, stat='density', kde=True, ax=ax)\n",
    "            ax.title.set_text(f'{m} {tool1} vs {tool2} l>10k')\n",
    "            k += 1\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f'{PATH}/figures/hist_length_{m}.png', bbox_inches='tight', dpi=300) \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print('Density distribution')\n",
    "for m in MODIFICATIONS:\n",
    "    print(m)\n",
    "    ts = df_no.loc[np.logical_and(df_no['modification'] == m, \n",
    "                                  ['pbp' in ty for ty in df_no['type']])]\n",
    "    plt.figure(figsize=(14, 4))\n",
    "    k = 1\n",
    "    for i in tqdm(range(len(tools))):\n",
    "        for j in range(i + 1, len(tools)):\n",
    "            tool1 = tools[i]\n",
    "            tool2 = tools[j]\n",
    "            if not (tool1 == 'SPAN' or tool2 == 'SPAN'):\n",
    "                continue\n",
    "\n",
    "            tst = ts[(ts['name'].str.startswith(f'{tool1} (')) | \n",
    "                     (ts['name'].str.startswith(f'{tool2} (')) | \n",
    "                     (ts['name'].str.contains(f'coverage {tool1} (', regex=False)) | \n",
    "                     (ts['name'].str.contains(f'coverage {tool2} (', regex=False)) | \n",
    "                     (ts['name'].str.contains(f'diff {tool1} - {tool2} (', regex=False)) | \n",
    "                     (ts['name'].str.contains(f'diff {tool2} - {tool1} (', regex=False))].copy()\n",
    "            tst['name'] = [re.sub(' \\(.*', '', n) for n in tst['name']]\n",
    "            # Plot             \n",
    "\n",
    "            tst.sort_values(by=['name', 'type'], inplace=True)\n",
    "            tst['value'].clip(upper=10.0, inplace=True)  # For visualizatioin\n",
    "            tst.rename(dict(value='RPKM'), axis=1, inplace=True)\n",
    "            ax = plt.subplot(1, 3, k)\n",
    "            k += 1\n",
    "            sns.histplot(data=tst, x='RPKM', hue='name', stat='density', kde=True, ax=ax)\n",
    "            ax.title.set_text(f'{m} {tool1} vs {tool2}')\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f'{PATH}/figures/hist_density_{m}.png', bbox_inches='tight', dpi=300) \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Peaks density vs length scatterplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "for m in MODIFICATIONS:\n",
    "    print(m)\n",
    "    ts = df_no.loc[\n",
    "        (df_no['modification'] == m) & ((df_no['type'] == 'signal pbp') | (df_no['type'] == 'length'))\n",
    "    ].copy()\n",
    "    ts['name'] = [re.sub(' \\(.*', '', n) for n in ts['name']]\n",
    "    plt.figure(figsize=(14, 4))\n",
    "    k = 1    \n",
    "    for i in tqdm(range(len(tools))):\n",
    "        for j in range(i + 1, len(tools)):\n",
    "            tool1 = tools[i]\n",
    "            tool2 = tools[j]\n",
    "            if not (tool1 == 'SPAN' or tool2 == 'SPAN'):\n",
    "                continue\n",
    "\n",
    "            # Use stratified sampling for performance reasons\n",
    "            tss = []            \n",
    "            for name in [tool1, tool2, f'coverage {tool1}', f'coverage {tool2}']:\n",
    "                tss.append(ts[ts['name'] == name].sample(100000))\n",
    "\n",
    "            # Plot\n",
    "            tst = pd.concat(tss)\n",
    "            t_len = tst[tst['type'] == 'length']\n",
    "            t_rpkm = tst[tst['type'] != 'length']\n",
    "            ax = plt.subplot(1, 3, k)\n",
    "            k += 1\n",
    "            sns.kdeplot(x=t_len['value'].clip(upper=20000).values,\n",
    "                        y=t_rpkm['value'].clip(upper=30.0).values,\n",
    "                        hue=t_rpkm['name'].values,\n",
    "                        fill=True, alpha=0.4, ax=ax)\n",
    "            ax.title.set_text(f'{m} {tool1} vs {tool2}')\n",
    "            ax.set_xlabel('Length')\n",
    "            ax.set_ylabel('RPKM')\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f'{PATH}/figures/density_vs_len_{m}.png', bbox_inches='tight', dpi=300) \n",
    "    plt.show"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
